Fix http/tests/security/cross-frame-access-call.html with site isolation enabled

rwlbuis opened the pull request:
#### 832ae665098cfe68dc899a7dfc5a795cde5ae3dc
<pre>
Fix http/tests/security/cross-frame-access-call.html with site isolation enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=303095">https://bugs.webkit.org/show_bug.cgi?id=303095</a>

Reviewed by Chris Dumez.

Fix a casting problem with site-isolation where the dynamicDowncast
to LocalDOMWindow even works for RemoteDOMWindow, by correcting the
type casts for DOMWindow/LocalDOMWindow given an EventTarget.

* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::IDLOperation&lt;JSEventTarget&gt;::call):
* Source/WebCore/page/DOMWindow.h:
(isType):
* Source/WebCore/page/LocalDOMWindow.h:
(isType):

Canonical link: <a href="https://commits.webkit.org/304046@main">https://commits.webkit.org/304046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52a9e15dea463b7815d8fe115e2dabb13c2f5478

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134214 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6725 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141798 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86274 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136084 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6589 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102632 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69910 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137161 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120333 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5000 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2579 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1610 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38471 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144441 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6395 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39051 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111013 "Found 1 new test failure: fast/scrolling/rtl-scrollbars-sticky-document.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6482 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5386 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111259 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28255 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4803 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116604 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60166 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6447 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34816 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6294 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69913 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6535 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/js/JSEventTargetCustom.h:
```diff
@@ -71,9 +73,12 @@ template<> class IDLOperation<JSEventTarget> {
-        auto& wrapped = thisObject.wrapped();
-        if (auto window = dynamicDowncast<LocalDOMWindow>(wrapped)) {
-            if (!window->frame() || !BindingSecurity::shouldAllowAccessToDOMWindow(&lexicalGlobalObject, *window, ThrowSecurityError))
+        if (auto *domWindow = thisObject.toJSDOMWindow()) {
```
Can you paste the crash on the bug? I'm having trouble understanding the fix. The cast was already a `dynamicDowncast` and `dynamicDowncast` is never supposed to crash so I am confused.

rwlbuis replied:
> Can you paste the crash on the bug? I'm having trouble understanding the fix. The cast was already a `dynamicDowncast` and `dynamicDowncast` is never supposed to crash so I am confused.

In fact my commit message may have been misleading and there was no crash in the `dynamicDowncast` as such, but as a result of the cast, i.e. the cast result window crashed when calling frame method on it.

Mainly though I did not understand the problem at the time, I debugged again today, and it seems to me the problem is that even though the dynamicDowncast was to LocalDOMWindow it really is dealing with a RemoteDOMWindow, where the script execution context/document will be null. To workaround this in my latest upload I call BindingSecurity::shouldAllowAccessToDOMWindow first, so window->frame will not be executed if there is a remote frame.

Maybe you can let me know if there is a way to fix the dynamicDowncast<LocalDOMWindow> part, I don't have much experience with dynamicDowncast but it seems to me SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::LocalDOMWindow) is very (too?) permissive by having the EventTarget overload. OTOH that line likely is not there by accident...

cdumez replied:
Yes, I suspect you're right and we need to fix the downcast traits. Something like this:
```diff
diff --git a/Source/WebCore/page/DOMWindow.h b/Source/WebCore/page/DOMWindow.h
index e383850011a7..714e04db5706 100644
--- a/Source/WebCore/page/DOMWindow.h
+++ b/Source/WebCore/page/DOMWindow.h
@@ -253,3 +253,7 @@ private:
 WebCoreOpaqueRoot root(DOMWindow*);
 
 } // namespace WebCore
+
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::DOMWindow)
+    static bool isType(const WebCore::EventTarget& target) { return target.eventTargetInterface() == WebCore::EventTargetInterfaceType::DOMWindow; }
+SPECIALIZE_TYPE_TRAITS_END()
diff --git a/Source/WebCore/page/LocalDOMWindow.h b/Source/WebCore/page/LocalDOMWindow.h
index 2413d546a573..8f614d50c916 100644
--- a/Source/WebCore/page/LocalDOMWindow.h
+++ b/Source/WebCore/page/LocalDOMWindow.h
@@ -530,5 +530,5 @@ inline String LocalDOMWindow::status() const
 
 SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::LocalDOMWindow)
     static bool isType(const WebCore::DOMWindow& window) { return window.isLocalDOMWindow(); }
-    static bool isType(const WebCore::EventTarget& target) { return target.eventTargetInterface() == WebCore::EventTargetInterfaceType::DOMWindow; }
+    static bool isType(const WebCore::EventTarget& target) { return is<DOMWindow>(target) && is<LocalDOMWindow>(downcast<DOMWindow>(target)); }
 SPECIALIZE_TYPE_TRAITS_END()
```

rwlbuis replied:
Thanks, that does indeed fix the LocalDOMWindow dynamicCast.

rwlbuis committed:
Fix http/tests/security/cross-frame-access-call.html with site isolation enabled
https://bugs.webkit.org/show_bug.cgi?id=303095

Reviewed by Chris Dumez.

Fix a casting problem with site-isolation where the dynamicDowncast
to LocalDOMWindow even works for RemoteDOMWindow, by correcting the
type casts for DOMWindow/LocalDOMWindow given an EventTarget.

* LayoutTests/platform/ios-site-isolation/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::IDLOperation<JSEventTarget>::call):
* Source/WebCore/page/DOMWindow.h:
(isType):
* Source/WebCore/page/LocalDOMWindow.h:
(isType):

Canonical link: https://commits.webkit.org/304046@main
